src/v/cluster/archival_metadata_stm.cc

```diff
@@ -599,6 +599,15 @@ archival_metadata_stm::archival_metadata_stm(
   , _cloud_storage_api(remote)
   , _feature_table(ft) {}
 
+archival_metadata_stm::~archival_metadata_stm() {
```

Comment:
I wondering if it makes sense to instead wait for the future in the `stop()` method? This looks a little fishy because it makes me wonder if there's some lifetime issue where something is still waiting on this future in the background, and then the stm gets destructed from underneath it.

I guess that wouldn't happen since the only waiter is sync(), which we'd always co_await for.

Reply:
The potential issue you mention is exactly the reason I decided to put it in the destructor rather than in stop. If the object is being destructed then there "can't" be anything depending on its state anymore. If anything does, they have bigger issues...

With stop, it is not as clear whether it is ok to "forget" this value.

Reply:
Yeah I think we'd only run into trouble if we were using raw futures instead of coroutines in the calls to sync. I think this is fine.